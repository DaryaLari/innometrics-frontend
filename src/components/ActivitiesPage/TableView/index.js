import React from 'react'
import { connect } from 'react-redux'
import _ from 'lodash'
import { getActivities } from '../../../helpers/selectors'
import styles from './style.css'

class TableView extends React.Component {
  state = {
    sortKey: null,
    order: null // ['asc' | 'desc']
  }
  sortBy = (key) => {
    if(this.state.order === null || this.state.order === 'desc'){
      this.setState({sortKey: key, order: 'asc'})
    }
    else{
      this.setState({sortKey: key, order: 'desc'})
    }
  }
  getSortedActivities = () => {
    const key = this.state.key
    const order = this.state.order
    if(key === null) {
      return this.props.activities
    }
    let activities = this.props.activities
    let modifiedKey = this.state.sortKey

    // if sortBy time, then convert value to data
    if(key === 'start_time' || key === 'end_time'){
      modifiedKey = (a) => Date.parse(a[key])
    }

    activities = _.sortBy(activities, modifiedKey)
    if(order === 'desc'){
      _.reverse(activities)
    }
    return activities
  }
  orderSign = (key) => {
    if(this.state.sortKey !== key){
      return ''
    }
    if(this.state.order === 'asc'){
      return <i className={`${'material-icons'} ${styles.sortDirection}`}>arrow_drop_up</i>
    }
    return <i className={`${'material-icons'} ${styles.sortDirection}`}>arrow_drop_down</i>
  }
  render(){
    let activities = this.getSortedActivities()
    return (
      <div className={styles.tableView}>
        <div className={styles.table}>
          <div className={styles.headerRow}>
            <div className={styles.cell}
                onClick={() => this.sortBy('start_time')}
            >
              Start time
              {this.orderSign('start_time')}
            </div>
            <div className={styles.cell}
                onClick={() => this.sortBy('end_time')}
            >
              End time
              {this.orderSign('end_time')}
            </div>
            <div className={styles.cell}
                onClick={() => this.sortBy('executable_name')}
            >
              File name
              {this.orderSign('executable_name')}
            </div>
          </div>
          {activities.map(a => (
            <div key={a._id}
                className={styles.row}
            >
              <div className={styles.cell}>{a.start_time}</div>
              <div className={styles.cell}>{a.end_time}</div>
              <div className={styles.cell} title={a.executable_name}>{a.executable_name}</div>
            </div>
          ))}
        </div>
      </div>
    )
  }
}

const ConnectedTableView = connect(
  (state) => ({
    activities: getActivities(state)
  })
)(TableView)

export default ConnectedTableView