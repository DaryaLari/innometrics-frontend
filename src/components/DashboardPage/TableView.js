import React from "react";
import styles from "./style.css";

const TableView = ({activities}) => {
  return (
    <div className={styles.tableView}>
      <table>
        <thead>
        <tr>
          <th>id</th>
          <th>user</th>
          <th>browser_title</th>
          <th>browser_url</th>
          <th>start_time</th>
          <th>end_time</th>
          <th>ip_address</th>
          <th>mac_address</th>
          <th>executable_name</th>
        </tr>
        </thead>
        <tbody>
        {activities.map(a => (
          <tr key={a._id}>
            <td>{a._id}</td>
            <td>{a.user}</td>
            <td>{a.browser_title}</td>
            <td>{a.browser_url}</td>
            <td>{a.start_time}</td>
            <td>{a.end_time}</td>
            <td>{a.ip_address}</td>
            <td>{a.mac_address}</td>
            <td>{a.executable_name}</td>
          </tr>
        ))}
        </tbody>
      </table>
    </div>
  )
}

export default TableView;