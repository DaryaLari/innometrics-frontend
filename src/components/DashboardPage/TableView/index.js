import React from "react";
import { connect } from 'react-redux'
import _ from 'lodash'
import { getActivities } from '../../../helpers/selectors'
import styles from "./style.css";

class TableView extends React.Component {
  state = {
    sortKey: null,
    order: null
  }
  sortBy = (key) => {
    if(this.state.order === null || this.state.order === 'desc'){
      this.setState({sortKey: key, order: 'asc'})
    }
    else{
      this.setState({sortKey: key, order: 'desc'})
    }
  }
  /*
  * @param { string } key - key of activities, by which the array should be sorted
  * @param { ['asc' | 'desc'] } order - for [ ascending or descending ] sorting order
  * */
  sortActivities = (key=this.state.key, order=this.state.order) => {
    if(key === null) {
      return this.props.activities
    }
    let activities = this.props.activities
    let modifiedKey = this.state.sortKey

    // if sortBy time, then convert value to data
    if(key === 'start_time' || key === 'end_time'){
      modifiedKey = (a) => Date.parse(a[key])
    }

    activities = _.sortBy(activities, modifiedKey)
    if(order === 'desc'){
      _.reverse(activities)
    }
    return activities
  }
  orderSign = (key) => {
    if(this.state.sortKey !== key){
      return ""
    }
    if(this.state.order === 'asc'){
      return <i className={`${"material-icons"} ${styles.sortDirection}`}>arrow_drop_up</i>
    }
    return <i className={`${"material-icons"} ${styles.sortDirection}`}>arrow_drop_down</i>
  }
  render(){
    let activities = this.sortActivities()
    return (
      <div className={styles.tableView}>
        <table className={styles.table}>
          <thead>
          <tr className={styles.row}>
            <th className={styles.cell}
                onClick={() => this.sortBy('start_time')}
            >
              Start time
              {this.orderSign('start_time')}
            </th>
            <th className={styles.cell}
                onClick={() => this.sortBy('end_time')}
            >
              End time
              {this.orderSign('end_time')}
            </th>
            <th className={styles.cell}
                onClick={() => this.sortBy('executable_name')}
            >
              File name
              {this.orderSign('executable_name')}
            </th>
          </tr>
          </thead>
          <tbody>
          {activities.map(a => (
            <tr key={a._id}
                className={`${styles.row} ${this.props.selectedActivity === a.executable_name && styles.selectedActivity}`}
            >
              <td className={styles.cell}>{a.start_time}</td>
              <td className={styles.cell}>{a.end_time}</td>
              <td className={styles.cell} title={a.executable_name}>{a.executable_name}</td>
            </tr>
          ))}
          </tbody>
        </table>
      </div>
    )
  }
}

const ConnectedTableView = connect(
  (state) => ({
    activities: getActivities(state)
  })
)(TableView)

export default ConnectedTableView;