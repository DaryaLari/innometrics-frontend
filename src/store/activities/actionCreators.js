import {TYPES as ACTIVITIES_TYPES} from './actionTypes'
import {getRequest} from '../../helpers/api'

export const getActivitiesRequest = () => (dispatch, getState) => {
  dispatch({type: ACTIVITIES_TYPES.GET_ACTIVITIES_REQUEST})

  getRequest('/activity', {offset: 0, amount_to_return: 100}, true)
      .then((result) => {
        dispatch({
          type: ACTIVITIES_TYPES.GET_ACTIVITIES_SUCCESS,
          payload: {
            activities: result.data.activities,
            offset: 0,
            amount_to_return: 100
          }
        })
      })
      .catch((error) => {
        dispatch({type: ACTIVITIES_TYPES.GET_ACTIVITIES_FAILURE, error: error.data.message})
      })
}