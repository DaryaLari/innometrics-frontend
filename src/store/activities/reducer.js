import {TYPES as ACTIVITIES_TYPES} from './actionTypes'
import {TYPES as USER_TYPES} from '../user/actionTypes'

const initialState = {
  activities: [],
  offset: 0,
  amount_to_return: 0,

  activeRequest: false,
  failed: false,
  error: null
}

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ACTIVITIES_TYPES.INIT_ACTIVITIES:
    case USER_TYPES.LOGIN_SUCCESS:
    case USER_TYPES.LOGOUT_SUCCESS:
    case USER_TYPES.LOGOUT_FAILURE:
      return initialState


    case ACTIVITIES_TYPES.GET_ACTIVITIES_REQUEST:
      return {
        ...state,
        activeRequest: true,
        failed: false,
        error: null
      }

    case ACTIVITIES_TYPES.GET_ACTIVITIES_SUCCESS:
      return {
        activities: action.payload.activities,
        offset: action.payload.offset,
        amount_to_return: action.payload.amount_to_return,
        activeRequest: false,
        failed: false,
        error: null
      }

    case ACTIVITIES_TYPES.GET_ACTIVITIES_FAILURE:
      return {
        ...state,
        activeRequest: false,
        failed: true,
        error: action.error
      }

    default:
      return state
  }
}