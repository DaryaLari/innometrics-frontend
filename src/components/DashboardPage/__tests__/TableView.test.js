import React from 'react'
import TestRenderer from 'react-test-renderer'
import {activities} from '../../../../___mocks___/activities'
import TableView from '../TableView'

describe('Test TableView component', () => {

  it('Test render TableView component', () => {

    const instance = TestRenderer.create(
      <TableView activities={activities} />
    )

    expect(instance.toJSON()).toMatchSnapshot()
  })

})