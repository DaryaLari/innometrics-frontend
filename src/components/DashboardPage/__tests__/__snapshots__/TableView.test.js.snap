// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test TableView component Test render TableView component 1`] = `
<div>
  <table>
    <thead>
      <tr>
        <th>
          start_time
        </th>
        <th>
          end_time
        </th>
        <th>
          executable_name
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          <span>
            2018-10-25 11:29:33
          </span>
        </td>
        <td>
          <span>
            2018-10-25 11:28:59
          </span>
        </td>
        <td>
          <span>
            file:///Users/&lt;username&gt;/Library/Developer/Xcode/DerivedData/InnometricsTransfer-ftpzsieukjkaktfvhcxnjxjwyjwi/Build/Products/Debug/InnometricsTransfer.app/Contents/MacOS/InnometricsTransfer
          </span>
        </td>
      </tr>
      <tr>
        <td>
          <span>
            2018-10-25 11:33:50
          </span>
        </td>
        <td>
          <span>
            2018-10-25 11:33:47
          </span>
        </td>
        <td>
          <span>
            file:///Users/&lt;username&gt;/Library/Developer/Xcode/DerivedData/InnometricsTransfer-ftpzsieukjkaktfvhcxnjxjwyjwi/Build/Products/Debug/InnometricsTransfer.app/Contents/MacOS/InnometricsTransfer
          </span>
        </td>
      </tr>
      <tr>
        <td>
          <span>
            2018-10-25 11:35:11
          </span>
        </td>
        <td>
          <span>
            2018-10-25 11:33:50
          </span>
        </td>
        <td>
          <span>
            file:///Applications/Xcode.app/Contents/MacOS/Xcode
          </span>
        </td>
      </tr>
      <tr>
        <td>
          <span>
            2018-10-25 11:33:45
          </span>
        </td>
        <td>
          <span>
            2018-10-25 11:33:40
          </span>
        </td>
        <td>
          <span>
            file:///Users/&lt;username&gt;/Library/Developer/Xcode/DerivedData/InnometricsTransfer-ftpzsieukjkaktfvhcxnjxjwyjwi/Build/Products/Debug/InnometricsTransfer.app/Contents/MacOS/InnometricsTransfer
          </span>
        </td>
      </tr>
      <tr>
        <td>
          <span>
            2018-10-25 11:29:56
          </span>
        </td>
        <td>
          <span>
            2018-10-25 11:29:54
          </span>
        </td>
        <td>
          <span>
            file:///Applications/Safari.app/Contents/MacOS/Safari
          </span>
        </td>
      </tr>
      <tr>
        <td>
          <span>
            2018-10-25 11:29:54
          </span>
        </td>
        <td>
          <span>
            2018-10-25 11:29:45
          </span>
        </td>
        <td>
          <span>
            file:///Users/&lt;username&gt;/Library/Developer/Xcode/DerivedData/InnometricsTransfer-ftpzsieukjkaktfvhcxnjxjwyjwi/Build/Products/Debug/InnometricsTransfer.app/Contents/MacOS/InnometricsTransfer
          </span>
        </td>
      </tr>
      <tr>
        <td>
          <span>
            2018-10-25 11:29:45
          </span>
        </td>
        <td>
          <span>
            2018-10-25 11:29:33
          </span>
        </td>
        <td>
          <span>
            file:///Applications/Safari.app/Contents/MacOS/Safari
          </span>
        </td>
      </tr>
      <tr>
        <td>
          <span>
            2018-10-25 11:30:00
          </span>
        </td>
        <td>
          <span>
            2018-10-25 11:29:56
          </span>
        </td>
        <td>
          <span>
            file:///Applications/PyCharm.app/Contents/MacOS/pycharm
          </span>
        </td>
      </tr>
      <tr>
        <td>
          <span>
            2018-10-25 11:33:40
          </span>
        </td>
        <td>
          <span>
            2018-10-25 11:30:00
          </span>
        </td>
        <td>
          <span>
            file:///Applications/Telegram.app/Contents/MacOS/Telegram
          </span>
        </td>
      </tr>
      <tr>
        <td>
          <span>
            2018-10-25 11:33:47
          </span>
        </td>
        <td>
          <span>
            2018-10-25 11:33:45
          </span>
        </td>
        <td>
          <span>
            file:///Applications/Xcode.app/Contents/MacOS/Xcode
          </span>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
